Add Git Stage action to editor title (#187893)

gjsjohnmurray opened the pull request:
This PR closes #187893

For git users it adds a Stage button to the editor title of a file with unstaged changes.

![image](https://github.com/microsoft/vscode/assets/6726799/dc32d85a-7615-4baa-8481-c53cbea59fee)

To achieve this I first added to core a new context key `scmActiveResourceGroupMemberships` which is an array of string ids of the resource groups the active resource is present in. The built-in Git extension uses this to contribute its Stage command to the title of an editor with unstaged changes.

I had to work around a limitation of the where-clause's `in` operator, which can only search an array-type context key for a value that matches another context key, not a constant.

I also used fractionally-negative order suffixes on the added command `git.stage` and its already-existing cousin `git.openChange` so they are adjacent but don't precede any `editor/title/run` contribution (screenshot shows this coming from the vscode-mock-debug extension).

If any existing users are upset by this button appearing they can easily use its context menu to demote it to the `...` menu.

gjsjohnmurray committed:
Add Git Stage action to editor title (#187893)

gjsjohnmurray commented:
/assign @lszomoru

gjsjohnmurray commented:
Updated from main and ready for your consideration @lszomoru

gjsjohnmurray commented:
@lszomoru any thoughts yet about this PR, which I submitted last August in response to you tagging the associated issue as "help wanted"?

irvnriir commented:
I think that this requires ability to hide the actions from the More Actions menu too. Currently, hiding buttons on this toolbar puts them to the menu, that becomes an issue if custom Extension fills that menu with custom items.

gjsjohnmurray committed:
Fix `git.stage` when `"git.untrackedChanges": "separate"`

gjsjohnmurray committed:
Only show button when file has unstaged/untracked changes

gjsjohnmurray commented:
Rebased to `main` then force-pushed to resolve complex conflict.

Also fixed Command Palette `Git: Stage Changes` not working for an untracked file if `"git.untrackedChanges": "separate"`.

gjsjohnmurray committed:
Merge branch 'main' into do-187893

gjsjohnmurray committed:
Merge branch 'main' into do-187893

gjsjohnmurray committed:
Merge branch 'main' into do-187893

gjsjohnmurray committed:
Merge branch 'main' into do-187893

lszomoru commented:
Sorry for not getting back to you on this until now. 

> To achieve this I first added to core a new context key scmActiveResourceGroupMemberships which is an array of string ids of the resource groups the active resource is present in. The built-in Git extension uses this to contribute its Stage command to the title of an editor with unstaged changes.

This is problematic from a performance perspective since you can have a large number of changed resources and we do not want push all this information from the git extension to core as a context key. We already have commands contributed by the git extension `git.stageChanges` and `git.stageFile` so we should find a way to surface these in the editor toolbar. We do have new API on core that "attaches" diff information to a TextEditor so maybe that is something that we could use. Let me think about it.

gjsjohnmurray commented:
I'm not clear where you are seeing my PR causing the Git extension to push a lot of data into core as a context key. Its constructor sets context keys for when-clauses to use with the `in` operator. The other new context key is maintained core-side.

lszomoru commented:
I wonder why this could not be implemented exclusively in the vscode.git extension:
- listen to `onDidChangeActiveTextEditor`, and set a context key if the resource is part of the working tree group
- use that context key to add the command to the editor title action bar

lszomoru commented:
Closing this in favour of https://github.com/microsoft/vscode/pull/237257.
